### PR TITLE
Reworked storage, reading and processing of auditer messages with JSON source

### PR DESCRIPTION
**Pull request by zlamalp**
DB
- Structure of table to store auditer messages has changed. Because
  Oracle doesn't support simple setting value of sequence, we couldn't
  switch code to using new table "auditer_log_json" prepared before.
  Instead we keep original name "auditer_log" and its sequence.
  Content is switched by renaming tables. We omit dropping original table
  in changelog, since it contains too much data and could be slow.
- Fixed init sql script for oracle with wrong order of column params.
- DB version bumped to 3.1.55

BASE
- AuditEvent object is now JSON type aware, so we don't have to manually
  read its class (name param) and construct it with class specification.
  Jackson does it for us automatically. Unused methods for this
  were removed.
- AuditMessage now contains AuditEvent instead of simple string message.
  It is still mainly used to read audit events/messages data.
- Removed compare methods in AuditMessage (object was not comparable).
- Added getUIMessage() in AuditMessage so we can display original
  message in HR format (original content is still parsable). It is used
  in GUI, but ignored on input of our API.
- Removed unused JsonSerializerGWT (GUI uses JsonSerializerJSONP).
- Updated JsonDeserializer (part of rpclib package) to reflect
  recent changes in same deserializer in rpc package (outer api).
  It is necessary for our old LDAPc to correctly parse AuditMessage.
- Updated test data and schema.

CORE
- Updated DB changelog.
- Logic for reading stored messages and manage consumers was moved
  from Auditer to standard core managers implementation (entry-bl-impl).
  New AuditMessagesManagerImpl was created and incorporated to bl/impl
  for this purpose.
- Storing new messages to DB is for now still kept within Auditer.
- Removed all logic which returned auditer messages as strings.
- Read clob as string for oracle (driver handles it), removed
  unnecessary long code for it.
- Updated tests, removed some ignored and commented tests.

LDAPC / LDAPC-INITIALIZER / LDAPC-ADA
- Fixes for handling new version of AuditerMessage object.
- We still parse original content of AuditerMessage.

GUI
- Updates to AuditMessage model.

OTHER
- Removed unused API and logic for AuditerMessages.
- Fixed and rewritten RPC docs for AuditMessagesManager.
- Added state changing check to API.
- Removed unused code from GUI.
- Fixed and updated javadocs.

Current state is, that engine/dispatcher and notifications consumes
AuditEvents, since they don't care about message ID. Old and new LDAPc
consumes AuditMessage, since they care about message ID. We now read
all data from JSON format (in table) and provide original "strings"
for parsing purpose still used in both LDAPCs.

Old LDAPc was locally manually tested and works. New LDAPc consumes
events directly from db (java) and works too.

### CODE REVIEW

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuditerMessage.java [hunk 1]
```diff
@@ -3,26 +3,47 @@
 import cz.metacentrum.perun.audit.events.AuditEvent;
 import cz.metacentrum.perun.core.api.PerunSession;
 
+/**
+ * Wrapper for AuditEvent which associated it with origination user session.
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuditerMessage.java]: *originating

File: perun-core/src/main/java/cz/metacentrum/perun/core/api/AuditMessagesManager.java [hunk 2]
```diff
@@ -11,87 +11,59 @@
...
 	 */
 	List<AuditMessage> getMessages(PerunSession perunSession, int count) throws InternalErrorException;
 
 	/**
-	 * Return less than count or equals to count messages from audit's logs.
-	 *
-	 * <b>IMPORTANT:</b> This variant do not guarantee returning just count of messages!
-	 * Return messages by Id from max_id to max_id-count (can be less then count messages).
+	 * Returns messages from audit log where param 'count' is applied to message IDs starting with current max_id.
+	 * It returns messages by their IDs from max_id to max_id-count (can be less than count messages).
```
[the remark this answers is not in the document]
Comment (zlamalp) [reply perun-core/src/main/java/cz/metacentrum/perun/core/api/AuditMessagesManager.java]: Hopefully fixed.

File: perun-core/src/main/java/cz/metacentrum/perun/core/bl/AuditMessagesManagerBl.java [hunk 3]
```diff
@@ -16,121 +16,87 @@
...
 	Map<String, Integer> getAllAuditerConsumers(PerunSession perunSession) throws InternalErrorException;
 
 	/**
-	 * Get id of last message from auditer_log.
+	 * Get ID of last message in auditer log (max_id).
 	 *
-	 * @return last message id
-	 * @throws InternalErrorException
+	 * @param perunSession perun session
+	 * @return Last message id
```
[the remark this answers is not in the document]
Comment (zlamalp) [reply perun-core/src/main/java/cz/metacentrum/perun/core/bl/AuditMessagesManagerBl.java]: I returned all throws declarations.

File: perun-core/src/main/java/cz/metacentrum/perun/core/bl/AuditMessagesManagerBl.java [hunk 4]
```diff
@@ -16,121 +16,87 @@
...
+	void setLastProcessedId(PerunSession perunSession, String consumerName, int lastProcessedId) throws InternalErrorException;
 
 	/**
-	 * Get number of messages in auditer log.
+	 * Get count of all messages in auditer log.
 	 *
-	 * @param perunSession
-	 * @return number of messages in auditer log
-	 * @throws InternalErrorException
+	 * @return Count of all messages in auditer log
```
[the remark this answers is not in the document]
Comment (zlamalp) [reply perun-core/src/main/java/cz/metacentrum/perun/core/bl/AuditMessagesManagerBl.java]: I returned all throws and param declarations to the javadoc.

File: perun-core/src/main/java/cz/metacentrum/perun/core/entry/AuditMessagesManagerEntry.java [hunk 5]
```diff
@@ -125,22 +86,22 @@ public void log(PerunSession perunSession, String message) throws InternalErrorE
...
+		getAuditMessagesManagerBl().setLastProcessedId(perunSession, consumerName, lastProcessedId);
 	}
 
 	@Override
-	public int getAuditerMessagesCount(PerunSession perunSession) throws InternalErrorException {
-		Utils.checkPerunSession(perunSession);
+	public int getAuditerMessagesCount(PerunSession perunSession) throws InternalErrorException, PrivilegeException {
+		if (!AuthzResolver.isAuthorized(perunSession, Role.PERUNADMIN)) {
+			throw new PrivilegeException(perunSession, "getAuditerMessagesCount");
+		}
```
[the remark this answers is not in the document]
Comment (zlamalp) [reply perun-core/src/main/java/cz/metacentrum/perun/core/entry/AuditMessagesManagerEntry.java]: I removed authorization for it.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuditMessagesManagerImpl.java [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+		mapper.getDeserializationConfig().addMixInAnnotations(Attribute.class, JsonDeserializer.AttributeMixIn.class);
+		mapper.getDeserializationConfig().addMixInAnnotations(AttributeDefinition.class, JsonDeserializer.AttributeDefinitionMixIn.class);
+		mapper.getDeserializationConfig().addMixInAnnotations(User.class, JsonDeserializer.UserMixIn.class);
+		mapper.getDeserializationConfig().addMixInAnnotations(Member.class, JsonDeserializer.MemberMixIn.class);
+		mapper.getDeserializationConfig().addMixInAnnotations(PerunBean.class, JsonDeserializer.PerunBeanMixIn.class);
+		mapper.getDeserializationConfig().addMixInAnnotations(Candidate.class, JsonDeserializer.CandidateMixIn.class);
+		mapper.getDeserializationConfig().addMixInAnnotations(PerunException.class, JsonDeserializer.PerunExceptionMixIn.class);
+		mapper.getDeserializationConfig().addMixInAnnotations(Destination.class, JsonDeserializer.DestinationMixIn.class);
+		mapper.getDeserializationConfig().addMixInAnnotations(Group.class, JsonDeserializer.GroupMixIn.class);
+		mapper.getDeserializationConfig().addMixInAnnotations(Group.class, JsonDeserializer.UserExtSourceMixIn.class);
```
[the remark this answers is not in the document]
Comment (zlamalp) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuditMessagesManagerImpl.java]: Fixed.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuditMessagesManagerImpl.java [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+
+	public AuditMessagesManagerImpl(DataSource perunPool) {
+		this.jdbc = new JdbcPerunTemplate(perunPool);
+	}
+
+	@Override
+	public List<AuditMessage> getMessages(PerunSession perunSession, int count) throws InternalErrorException {
+		try {
+			return jdbc.query("select " + auditMessageMappingSelectQuery + " from (select " + auditMessageMappingSelectQuery + Compatibility.getRowNumberOver() + " from auditer_log ORDER BY id Desc limit ?) "+Compatibility.getAsAlias("temp"),
+					AUDIT_MESSAGE_MAPPER, count);
```
[the remark this answers is not in the document]
Comment (zlamalp) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuditMessagesManagerImpl.java]: My bad, now it should work on both DBs using "rownumber" column.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuditMessagesManagerImpl.java [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+
+	@Override
+	public List<AuditEvent> pollConsumerEvents(PerunSession perunSession, String consumerName) throws InternalErrorException {
+		if (consumerName == null) throw new InternalErrorException("Auditer consumer doesn't exist.");
+
+		try {
+
+			if(jdbc.queryForInt("select count(*) from auditer_consumers where name=?", consumerName) != 1) {
+				throw new InternalErrorException("Auditer consumer doesn't exist.");
+			}
```
[the remark this answers is not in the document]
Comment (zlamalp) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuditMessagesManagerImpl.java]: Added method ``checkAuditerConsumerExists()`` and used where appropriate.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuditMessagesManagerImpl.java [hunk 9]
```diff
@@ -0,0 +1,293 @@
...
+
+			List<AuditEvent> eventList = new ArrayList<>();
+
+			int lastProcessedId = getLastProcessedId(consumerName);
+			int maxId = jdbc.queryForInt("select max(id) from auditer_log");
```
[the remark this answers is not in the document]
Comment (zlamalp) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuditMessagesManagerImpl.java]: Such method already exists -> ``getLastMessageId()``, I'm using it now for this purpose in both ``pollConsumerMessages/Events()`` methods.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuditMessagesManagerImpl.java [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+			if (maxId > lastProcessedId) {
+
+				eventList = jdbc.query("select " + auditMessageMappingSelectQuery + " from auditer_log where id > ? and id <= ? order by id", AUDIT_EVENT_MAPPER, lastProcessedId, maxId);
+
+				// we successfully de-serialized all messages, bump counter
+				jdbc.update("update auditer_consumers set last_processed_id=?, modified_at=" + Compatibility.getSysdate() + " where name=?", maxId, consumerName);
```
[the remark this answers is not in the document]
Comment (zlamalp) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuditMessagesManagerImpl.java]: Fixed in both methods.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuditerMessage.java (hunk 1 above)
Comment (zlamalp) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuditerMessage.java]: Fixed.

File: perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuditMessagesManagerImpl.java [hunk 11]
```diff
@@ -259,15 +253,21 @@ public void createAuditerConsumer(PerunSession perunSession, String consumerName
...
 			throw new InternalErrorException(ex);
 		}
+	}
 
-		return auditerConsumers;
+	@Override
+	public boolean checkAuditerConsumerExists(PerunSession session, String consumerName) throws InternalErrorException {
+		if (consumerName == null) throw new InternalErrorException("Auditer consumer doesn't exist.");
+		try {
+			return jdbc.queryForInt("select count(*) from auditer_consumers where name=?", consumerName) == 1;
```
Comment (stavamichal) [inline perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuditMessagesManagerImpl.java]: I am afraid there could be more than 1 consumer with such name. You should trow an exception if number of consumers is bigger than 1.

Comment (zlamalp) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuditMessagesManagerImpl.java]: There is unique constraint on the column, so we don't have to check on duplicate consumers in the code.

Comment (stavamichal) [reply perun-core/src/main/java/cz/metacentrum/perun/core/impl/AuditMessagesManagerImpl.java]: You are right.

**Pull request merged**